Fix for `‘__m128’ does not name a type` compile error.

lrdatacomfort opened the pull request:
Fixes ‘__m128’ complication error on QEMU "QEMU Virtual CPU version 2.5+" on x86_64:
```
src/nn/llamafile/sgemm.cpp:84:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
   84 | inline __m128 add(__m128 x, __m128 y) { return _mm_add_ps(x, y); }
      |        ^~~~~~
      |        __int128
src/nn/llamafile/sgemm.cpp:85:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
   85 | inline __m128 sub(__m128 x, __m128 y) { return _mm_sub_ps(x, y); }
      |        ^~~~~~
      |        __int128
src/nn/llamafile/sgemm.cpp:86:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
   86 | inline __m128 mul(__m128 x, __m128 y) { return _mm_mul_ps(x, y); }
      |        ^~~~~~
      |        __int128
src/nn/llamafile/sgemm.cpp:169:1: error: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++17’ [-Werror=c++17-extensions]
  169 | inline float hsum(__m128 x) {
      | ^~~~~~
src/nn/llamafile/sgemm.cpp:169:19: error: ‘__m128’ was not declared in this scope; did you mean ‘__int128’?
  169 | inline float hsum(__m128 x) {
      |                   ^~~~~~
      |                   __int128
src/nn/llamafile/sgemm.cpp:217:20: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
  217 | template <> inline __m128 load(const float *p) {
      |                    ^~~~~~
      |                    __int128
cc1plus: all warnings being treated as errors
make: *** [Makefile:55: llamafile-sgemm.o] Error 1
```

lrdatacomfort committed:
Fix for `‘__m128’ does not name a type` compile error.

geerlingguy commented:
Can you provide more context for this? Is this when running dllama in a specific scenario, or in all cases? I haven't run into this problem before.

lrdatacomfort commented:
Ah! I see i forgot to include that it's in distributed-llama. Sorry about that.
It's just when running your playbook without any modifications:
```bash
~$ git clone https://github.com/geerlingguy/beowulf-ai-cluster.git
Cloning into 'beowulf-ai-cluster'...
remote: Enumerating objects: 210, done.
remote: Counting objects: 100% (210/210), done.
remote: Compressing objects: 100% (148/148), done.
remote: Total 210 (delta 129), reused 127 (delta 58), pack-reused 0 (from 0)
Receiving objects: 100% (210/210), 414.55 KiB | 7.82 MiB/s, done.
Resolving deltas: 100% (129/129), done.
~$ cd beowulf-ai-cluster
~/beowulf-ai-cluster$ cp example.hosts.ini hosts.ini
~/beowulf-ai-cluster$ cp example.config.yml config.yml
~/beowulf-ai-cluster$ ansible-playbook main.yml

PLAY [Set up llama.cpp.] ********************************************************************************************************************************************************************************

TASK [Gathering Facts] **********************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [ansible.builtin.include_tasks] ********************************************************************************************************************************************************************
skipping: [127.0.0.1]

TASK [ansible.builtin.include_tasks] ********************************************************************************************************************************************************************
included: /home/user/beowulf-ai-cluster/dependencies/debian-based.yml for 127.0.0.1

TASK [Update apt cache.] ********************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Install dependencies.] ****************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Install extra packages.] **************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Ensure working directory exists.] *****************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Clone llama.cpp.] *********************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Create llama.cpp build directory.] ****************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Configure the build.] *****************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Build llama.cpp.] *********************************************************************************************************************************************************************************
ok: [127.0.0.1]

PLAY [Set up distributed-llama.] ************************************************************************************************************************************************************************

TASK [Gathering Facts] **********************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [ansible.builtin.include_tasks] ********************************************************************************************************************************************************************
skipping: [127.0.0.1]

TASK [ansible.builtin.include_tasks] ********************************************************************************************************************************************************************
included: /home/user/beowulf-ai-cluster/dependencies/debian-based.yml for 127.0.0.1

TASK [Update apt cache.] ********************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Install dependencies.] ****************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Ensure working directory exists.] *****************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Clone distributed-llama.] *************************************************************************************************************************************************************************
changed: [127.0.0.1]

TASK [Build distributed-llama.] *************************************************************************************************************************************************************************
[ERROR]: Task failed: Module failed: src/nn/llamafile/sgemm.cpp:84:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
   84 | inline __m128 add(__m128 x, __m128 y) { return _mm_add_ps(x, y); }
      |        ^~~~~~
      |        __int128
src/nn/llamafile/sgemm.cpp:85:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
   85 | inline __m128 sub(__m128 x, __m128 y) { return _mm_sub_ps(x, y); }
      |        ^~~~~~
      |        __int128
src/nn/llamafile/sgemm.cpp:86:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
   86 | inline __m128 mul(__m128 x, __m128 y) { return _mm_mul_ps(x, y); }
      |        ^~~~~~
      |        __int128
src/nn/llamafile/sgemm.cpp:169:1: error: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++17’ [-Werror=c++17-extensions]
  169 | inline float hsum(__m128 x) {
      | ^~~~~~
src/nn/llamafile/sgemm.cpp:169:19: error: ‘__m128’ was not declared in this scope; did you mean ‘__int128’?
  169 | inline float hsum(__m128 x) {
      |                   ^~~~~~
      |                   __int128
src/nn/llamafile/sgemm.cpp:217:20: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
  217 | template <> inline __m128 load(const float *p) {
      |                    ^~~~~~
      |                    __int128
cc1plus: all warnings being treated as errors
gmake: *** [Makefile:55: llamafile-sgemm.o] Fejl 1
Origin: /home/user/beowulf-ai-cluster/plays/dllama.yml:28:7

26       become: true
27
28     - name: Build distributed-llama.
         ^ column 7

failed: [127.0.0.1] (item=dllama) =>
    ansible_loop_var: item
    changed: false
    cmd: /usr/bin/gmake dllama
    item: dllama
    msg: |-
        src/nn/llamafile/sgemm.cpp:84:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           84 | inline __m128 add(__m128 x, __m128 y) { return _mm_add_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:85:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           85 | inline __m128 sub(__m128 x, __m128 y) { return _mm_sub_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:86:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           86 | inline __m128 mul(__m128 x, __m128 y) { return _mm_mul_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:169:1: error: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++17’ [-Werror=c++17-extensions]
          169 | inline float hsum(__m128 x) {
              | ^~~~~~
        src/nn/llamafile/sgemm.cpp:169:19: error: ‘__m128’ was not declared in this scope; did you mean ‘__int128’?
          169 | inline float hsum(__m128 x) {
              |                   ^~~~~~
              |                   __int128
        src/nn/llamafile/sgemm.cpp:217:20: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
          217 | template <> inline __m128 load(const float *p) {
              |                    ^~~~~~
              |                    __int128
        cc1plus: all warnings being treated as errors
        gmake: *** [Makefile:55: llamafile-sgemm.o] Fejl 1
    rc: 2
    stderr: |-
        src/nn/llamafile/sgemm.cpp:84:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           84 | inline __m128 add(__m128 x, __m128 y) { return _mm_add_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:85:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           85 | inline __m128 sub(__m128 x, __m128 y) { return _mm_sub_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:86:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           86 | inline __m128 mul(__m128 x, __m128 y) { return _mm_mul_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:169:1: error: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++17’ [-Werror=c++17-extensions]
          169 | inline float hsum(__m128 x) {
              | ^~~~~~
        src/nn/llamafile/sgemm.cpp:169:19: error: ‘__m128’ was not declared in this scope; did you mean ‘__int128’?
          169 | inline float hsum(__m128 x) {
              |                   ^~~~~~
              |                   __int128
        src/nn/llamafile/sgemm.cpp:217:20: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
          217 | template <> inline __m128 load(const float *p) {
              |                    ^~~~~~
              |                    __int128
        cc1plus: all warnings being treated as errors
        gmake: *** [Makefile:55: llamafile-sgemm.o] Fejl 1
    stderr_lines: <omitted>
    stdout: |-
        g++ -std=c++11 -Werror -Wformat -Werror=format-security  -march=native -mtune=native -O3 -DDLLAMA_VULKAN -c src/nn/nn-quants.cpp -o nn-quants.o
        g++ -std=c++11 -Werror -Wformat -Werror=format-security  -march=native -mtune=native -O3 -DDLLAMA_VULKAN -c src/nn/nn-core.cpp -o nn-core.o
        g++ -std=c++11 -Werror -Wformat -Werror=format-security  -march=native -mtune=native -O3 -DDLLAMA_VULKAN -c src/nn/nn-executor.cpp -o nn-executor.o
        g++ -std=c++11 -Werror -Wformat -Werror=format-security  -march=native -mtune=native -O3 -DDLLAMA_VULKAN -c src/nn/nn-network.cpp -o nn-network.o
        g++ -std=c++11 -Werror -Wformat -Werror=format-security  -march=native -mtune=native -O3 -DDLLAMA_VULKAN -c src/nn/llamafile/sgemm.cpp -o llamafile-sgemm.o
    stdout_lines: <omitted>
failed: [127.0.0.1] (item=dllama-api) =>
    ansible_loop_var: item
    changed: false
    cmd: /usr/bin/gmake dllama-api
    item: dllama-api
    msg: |-
        src/nn/llamafile/sgemm.cpp:84:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           84 | inline __m128 add(__m128 x, __m128 y) { return _mm_add_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:85:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           85 | inline __m128 sub(__m128 x, __m128 y) { return _mm_sub_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:86:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           86 | inline __m128 mul(__m128 x, __m128 y) { return _mm_mul_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:169:1: error: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++17’ [-Werror=c++17-extensions]
          169 | inline float hsum(__m128 x) {
              | ^~~~~~
        src/nn/llamafile/sgemm.cpp:169:19: error: ‘__m128’ was not declared in this scope; did you mean ‘__int128’?
          169 | inline float hsum(__m128 x) {
              |                   ^~~~~~
              |                   __int128
        src/nn/llamafile/sgemm.cpp:217:20: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
          217 | template <> inline __m128 load(const float *p) {
              |                    ^~~~~~
              |                    __int128
        cc1plus: all warnings being treated as errors
        gmake: *** [Makefile:55: llamafile-sgemm.o] Fejl 1
    rc: 2
    stderr: |-
        src/nn/llamafile/sgemm.cpp:84:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           84 | inline __m128 add(__m128 x, __m128 y) { return _mm_add_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:85:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           85 | inline __m128 sub(__m128 x, __m128 y) { return _mm_sub_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:86:8: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
           86 | inline __m128 mul(__m128 x, __m128 y) { return _mm_mul_ps(x, y); }
              |        ^~~~~~
              |        __int128
        src/nn/llamafile/sgemm.cpp:169:1: error: inline variables are only available with ‘-std=c++17’ or ‘-std=gnu++17’ [-Werror=c++17-extensions]
          169 | inline float hsum(__m128 x) {
              | ^~~~~~
        src/nn/llamafile/sgemm.cpp:169:19: error: ‘__m128’ was not declared in this scope; did you mean ‘__int128’?
          169 | inline float hsum(__m128 x) {
              |                   ^~~~~~
              |                   __int128
        src/nn/llamafile/sgemm.cpp:217:20: error: ‘__m128’ does not name a type; did you mean ‘__int128’?
          217 | template <> inline __m128 load(const float *p) {
              |                    ^~~~~~
              |                    __int128
        cc1plus: all warnings being treated as errors
        gmake: *** [Makefile:55: llamafile-sgemm.o] Fejl 1
    stderr_lines: <omitted>
    stdout: |-
        g++ -std=c++11 -Werror -Wformat -Werror=format-security  -march=native -mtune=native -O3 -DDLLAMA_VULKAN -c src/nn/llamafile/sgemm.cpp -o llamafile-sgemm.o
    stdout_lines: <omitted>

PLAY RECAP **********************************************************************************************************************************************************************************************
127.0.0.1                  : ok=16   changed=1    unreachable=0    failed=1    skipped=2    rescued=0    ignored=0

~/beowulf-ai-cluster$
```
